### PR TITLE
TF generate refactor - XLA sample

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR brings XLA to `sample`, in `generate`. Four important details before reviewing:
1. The diff has the changes of https://github.com/huggingface/transformers/pull/16704, review that PR first plz :) It fixes a test from `beam_search`. I will rebase as soon as the other PR gets merged (the changes were bundled to confirm that it passes all generate tests).
2. The body is mostly copy/paste from `greedy_search`;
3. The sample step was changed from the previous implementation -- if we want to seed sampling with XLA, we need to use the `stateless` functions;
4. The XLA sample tests do not compare all generated tokens to their non-XLA sample counterparts, due to the numerical instabilities discussed on Slack. We do compare the first tokens, which are the same.

Finally, tests have been run for the usual models (`gpt2`, `t5`, `rag`, `speech2text`, `encoder_decoder`, `vision_encoder_decoder`, `bart`). 

____________________________

I've also run a quick sanity check on GPU. Using GPT2+sample, on an Nvidia T4:
- eager TF: ~1.7s
- XLA TF: ~54ms (~22s compile time) :point_right: 31x speedup

### CODE REVIEW

File: tests/gpt2/test_modeling_tf_gpt2.py [hunk 1]
```diff
@@ -447,19 +447,6 @@ def test_model_from_pretrained(self):
...
-        input_ids = tf.convert_to_tensor([[464, 1893]], dtype=tf.int32)  # The president
-
-        # The president of the United States, and the president of the United Kingdom, have been in the White
-        # fmt: off
-        expected_output_ids = [464, 1893, 286, 262, 1578, 1829, 11, 290, 262, 1893, 286, 262, 1578, 7526, 11, 423, 587, 287, 262, 2635]
-        # fmt: on
-
-        output_ids = model.generate(input_ids, do_sample=False)
-        self.assertListEqual(output_ids[0].numpy().tolist(), expected_output_ids)
-
```
Comment (gante) [inline tests/gpt2/test_modeling_tf_gpt2.py:462]: This test was pretty much the same as `test_lm_generate_gpt2` (the only difference was the starting `input_ids`)

File: src/transformers/generation_tf_utils.py [hunk 2]
```diff
@@ -2220,97 +2232,155 @@ def sample(
...
+            next_tokens_scores = logits_warper(input_ids, next_tokens_scores)
 
             # sample
+            if seed is not None:
+                sample_seed = seed
+            else:
+                sample_seed = tf.cast(self.seed_generator.make_seeds(count=1)[:, 0], dtype=tf.int32)
             next_tokens = tf.squeeze(
-                tf.random.categorical(logits=next_token_scores, num_samples=1, dtype=tf.int32), axis=1
+                tf.random.stateless_categorical(
```
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:2309]: nice!

File: src/transformers/generation_tf_utils.py [hunk 3]
```diff
@@ -2119,6 +2127,7 @@ def sample(
         max_length: Optional[int] = None,
         pad_token_id: Optional[int] = None,
         eos_token_id: Optional[int] = None,
+        seed: Optional[int] = None,
```
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:2130]: ok for me!

File: src/transformers/generation_tf_utils.py [hunk 4]
```diff
@@ -1366,6 +1370,9 @@ def _generate(
             use_cache (`bool`, *optional*, defaults to `True`):
                 Whether or not the model should use the past last key/values attentions (if applicable to the model) to
                 speed up decoding.
+            seed (`List[int]`, *optional*):
```
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:1373]: Is it a list of ints or just int? Or both? Don't think the type hint is correct here

File: src/transformers/generation_tf_utils.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/transformers/generation_tf_utils.py:2130]: different type hint then above

File: tests/gpt2/test_modeling_tf_gpt2.py [hunk 5]
```diff
@@ -574,3 +561,16 @@ def test_lm_generate_gpt2_xla(self):
 
         output_ids = xla_generate(input_ids, do_sample=False)
         self.assertListEqual(output_ids[0].numpy().tolist(), expected_output_ids)
+
+    @slow
+    def test_lm_generate_gpt2_xla_sample(self):
```
Comment (patrickvonplaten) [inline tests/gpt2/test_modeling_tf_gpt2.py:566]: nice!

File: tests/gpt2/test_modeling_tf_gpt2.py [hunk 6]
```diff
@@ -574,3 +561,16 @@ def test_lm_generate_gpt2_xla(self):
...
+        model = TFGPT2LMHeadModel.from_pretrained("gpt2")
+        input_ids = tf.convert_to_tensor([[464, 3290]], dtype=tf.int32)  # The dog
+
+        # fmt: off
+        expected_output_ids = [464, 3290, 550, 284, 307, 4376, 287, 281, 4044, 1363, 329, 734, 812, 878, 852, 4376, 757, 329, 2267, 0]
+        # fmt: on
+        xla_generate = tf.function(model.generate, jit_compile=True)
+
+        output_ids = xla_generate(input_ids, do_sample=True, seed=[42, 0])
```
Comment (patrickvonplaten) [inline tests/gpt2/test_modeling_tf_gpt2.py:575]: Why is seed a list of ints ? Don't fully understand this

**Review by patrickvonplaten — APPROVED** [review]
Great! Awesome to see such a speed-up ! 

Looks good to me. The only thing that is not very intuitive to me is that `seed` is a list of integers - why is this? Is this normal in TF?

**Comment by gante** [comment]
@patrickvonplaten the `stateless` TF functions accept a `seed` argument that is a tuple of two integers 😅 Not very intuitive, I agree. They correspond to the `key` and `counter` used in the internal RNG algorithms ([source](https://www.tensorflow.org/api_docs/python/tf/random/Generator#from_key_counter)).

If you think it will be unintuitive for users, I can change it so that our `seed` argument corresponds to the `key` of the tuple, and fix the `counter` to `0`. For practical purposes, it should be the same thing.

**Comment by patrickvonplaten** [comment]
> @patrickvonplaten the `stateless` TF functions accept a `seed` argument that is a tuple of two integers sweat_smile Not very intuitive, I agree. They correspond to the `key` and `counter` used in the internal RNG algorithms ([source](https://www.tensorflow.org/api_docs/python/tf/random/Generator#from_key_counter)).
> 
> If you think it will be unintuitive for users, I can change it so that our `seed` argument corresponds to the `key` of the tuple (i.e. a single integer), and fix the `counter` to `0`. For practical purposes, it should be the same thing.

I see - ok maybe better to leave as is then to be aligned with TF

File: src/transformers/generation_tf_utils.py (hunk 3 above)
Comment (Rocketknight1) [inline src/transformers/generation_tf_utils.py:2130]: This is a list/tuple of two ints, right?
```suggestion
        seed: Optional[Sequence[int, int]] = None,
```

Comment (gante) [reply src/transformers/generation_tf_utils.py:2130]: yes, correct (adding Patrick's suggestions now)

File: src/transformers/generation_tf_utils.py (hunk 2 above)
Comment (Rocketknight1) [reply src/transformers/generation_tf_utils.py:2309]: If I understand this correctly, if the user passes a seed tuple, the same seed is used on every sampling for the entire generation run? That's not a problem, just making sure I got it right!

File: src/transformers/generation_tf_utils.py (hunk 3 above)
Comment (Rocketknight1) [reply src/transformers/generation_tf_utils.py]: You might have to `from typing import Sequence` at the top too, sorry about that!

File: tests/t5/test_modeling_tf_t5.py [hunk 7]
```diff
@@ -524,6 +524,35 @@ def test_greedy_generate(self):
...
+    def test_sample_xla_generate_simple(self):
+        model = TFT5ForConditionalGeneration.from_pretrained("t5-small")
+        tokenizer = T5Tokenizer.from_pretrained("t5-small")
+
+        sentence = "Translate English to German: Today is a beautiful day."
+        input_ids = tokenizer(sentence, return_tensors="tf", padding=True).input_ids
+        # XLA reorder ops, which causes operations like FP matmul to have slightly different results, causing
+        # divergences in generate -- especially with sampling.
+        expected_output_string = ["Heute ist ein schöner Tag."]
+        expected_output_string_xla = ["Heute ist ein schöne Tage."]
```
Comment (Rocketknight1) [inline tests/t5/test_modeling_tf_t5.py:537]: Small comment but "schöner" is correct here - I hope it's just sampling issues and not a sign of a bug that the XLA version gets it wrong!

Comment (Rocketknight1) [reply tests/t5/test_modeling_tf_t5.py:537]: Could you try, as a kind of stupid manual test, asking it to translate a batch of sentences from English to Portuguese and make sure that even if they're different, the quality of the XLA ones is similar to the quality of the manual ones? Numerical bugs can be very annoying to catch, but if the quality is similar then that would make me confident that the XLA implementation is not worse.

**Review by Rocketknight1 — APPROVED** [review]
Overall this looks great! I have a couple of very nitpicky nitpicks, mostly because I had a bug in my XLA implementation of greedy that I didn't notice for a long time because it only degraded the quality of sampling, so now I'm paranoid about catching bugs like that by checking output quality.

**Comment by gante** [comment]
While running tests for T5 (as suggested by @Rocketknight1), I found out that our XLA code is not behaving properly for T5, for both `sample` and `greedy_search`. Because the problem is not exclusive to `sample`, I'm merging this PR and fixing the issue in a future one.

(example)
image
